torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -1652,3 +1657,54 @@ def test_function_not_on_callee(self):
                 AttributeError, "RPC pickler does not serialize"
             ):
                 rpc.rpc_sync(callee_worker, foo_add, args=())
+
+@unittest.skipIf(
+    not torch._six.PY3,
+    "Pytorch distributed autograd package " "does not support python2",
+)
+class FaultyAgentRpcTest(RpcAgentTestFixture):
+    @dist_init(faulty_messages=["RREF_USER_DELETE"])
+    def test_pass_local_rrefs(self):
+        n = self.rank + 1
+        dst_rank = n % self.world_size
+        dst_worker = "worker{}".format(dst_rank)
+
+        rref = RRef(40)
+        self.assertEqual(
+            rpc.rpc_sync(dst_worker, add_rref_to_value, args=(rref, 50)), 90
+        )
+        self.assertEqual(
+            rpc.rpc_async(dst_worker, add_rref_to_value, args=(rref, 50)).wait(), 90
+        )
+        self.assertEqual(
+            rpc.remote(dst_worker, add_rref_to_value, args=(rref, 50)).to_here(), 90
+        )
+        wait_until_pending_forks_flushed()
+
+    @dist_init(faulty_messages=["RREF_CHILD_ACCEPT"])
+    def test_return_local_rrefs(self):
```

Comment:
Can we have just one test instead of two and mark all the RRef messages as faulty in that test? The test could be simple and only do this:

1. A creates an RRef on B.
2. A shared the RRef to C by calling a function on C with the RRef as an argument.
3. Finally, verify all RRefs have been cleaned up by calling `wait_until_pending_forks_flushed`.

In `wait_until_pending_forks_flushed` you can extend it to wait for num_forks and num_pending_users to hit 0. In addition to this if possible we could try explicitly deleting the RRef and also ensuring `num_owner_rrefs` hits 0.